src/compact_live.rs

```diff
@@ -0,0 +1,454 @@
+                k_value,
+                encrypted_value
+            )?;
+            if noise.contains(uid) {
+                // Noise entries are simply re-encrypted.
+                noisy_entry_table.insert(uid.clone(), entry_table_value);
+            } else {
+                // Compact chains associated to the other entries.
+                let chain = noisy_chain_values.get(uid).ok_or(Error::<CustomError>::CryptoError("No matching UID in chains.".to_string()))?;
+                let mut compacted_chain = HashMap::with_capacity(chain.len());
+                for value in chain {
+                    match value {
+                        IndexedValue::Location(location) => {
+                            if noisy_remaining_locations.contains(location) {
+                                // Add remaining locations back. All deletions should have been
+                                // filtered by the fetch.
+                                compacted_chain.insert(
+                                    IndexedValue::from(location.clone()),
+                                    BlockType::Addition,
+                                );
+                            }
+                        }
+                        IndexedValue::NextKeyword(keyword) => {
+                            // TODO (TBZ): deal with obsolete `NextWord`s. For now just add the
+                            // `NextWord` again.
+                            compacted_chain
+                                .insert(IndexedValue::from(keyword.clone()), BlockType::Addition);
+                        }
+                    };
+                }
+                compacted_chains.insert(uid.clone(), (entry_table_value.keyword_hash, compacted_chain));
+                cache.insert(entry_table_value.keyword_hash, uid.clone());
+            }
+        }
+
+        let compacted_chain_table = noisy_entry_table.upsert::<
+            CHAIN_TABLE_WIDTH,
+            BLOCK_LENGTH,
+            KMAC_KEY_LENGTH,
+            DEM_KEY_LENGTH,
+            KmacKey,
+            DemScheme
+        >(rng, &compacted_chains)?;
+
+        Ok((noisy_entry_table, compacted_chain_table))
+    }
+
+    /// Live compact the given UIDs.
+    ///
+    /// - `rng`         : secure random number generator
+    /// - `k_value`     : DEM key used to decrypt the Entry Table
+    /// - `noisy_uids`  : UIDs to compact (contains noise)
+    /// - `noise`       : Entry Table UIDs used as noise
+    async fn live_compact_uids(
+        &mut self,
+        rng: &mut impl CryptoRngCore,
+        k_value: &DemScheme::Key,
+        noisy_uids: Vec<Uid<UID_LENGTH>>,
+        noise: &HashSet<Uid<UID_LENGTH>>,
+    ) -> Result<(), Error<CustomError>> {
+        // Fetch all index entries (noisy).
+        let mut noisy_encrypted_entry_table = self
+            .fetch_entry_table(&HashSet::from_iter(noisy_uids.into_iter()))
+            .await?;
```

Comment:
I think you should fetch all the entries here (required uids and the ones from the noise) because otherwise you leak which keywords will be updated.

Reply:
My bad, I didn't see that `noisy_uids` contains the noise one… Not sure about it, it's wasn't clear for me…

Reply:
I think it's better to have only the real uids to compact inside this Vec, and merge the two sets when calling `fetch_entry_table`

Reply:
I prefer something like `noise_and_target_uids` / `noise_uids` / `target_uids` in that case. `noisy` could be easily interpreted as "only noise" or "in random order" or "fetched noisly" like in `noisy_fetch_chains` in search operation.

Reply:
Changed to `mixt_uids`. The merging of the target and noise sets can be implemented in the `for` loop in a later time.